cmd/vspd/spentticket.go

```diff
@@ -133,42 +134,40 @@ func (v *vspd) findSpentTickets(toCheck database.TicketList, startHeight int64)
 			return nil, 0, err
 		}
 
-		var iBlock *wire.MsgBlock
-	outer:
-		for ticketHash, ticket := range tickets {
-			if filter.Match(key, ticket.pkScript) {
-				// Filter match means the ticket is likely spent in block. Get
-				// the full block to confirm.
-				if iBlock == nil {
-					iBlock, err = dcrdClient.GetBlock(iHash)
-					if err != nil {
-						return nil, 0, err
-					}
-				}
+		if !filter.MatchAny(key, scripts) {
+			// No tickets are spent in this block, continue to the next one.
+			continue
+		}
 
-				// The regular transaction tree does not need to be checked
-				// because tickets can only be spent by vote or revoke
-				// transactions which are always in the stake tree.
-				for _, blkTx := range iBlock.STransactions {
-					if !txSpendsTicket(blkTx, ticketHash) {
-						continue
-					}
-
-					// Confirmed - ticket is spent in block.
-
-					spent = append(spent, spentTicket{
-						dbTicket:     ticket.dbTicket,
-						expiryHeight: ticket.dbTicket.PurchaseHeight + int64(params.TicketMaturity) + int64(params.TicketExpiry),
-						heightSpent:  iHeight,
-						spendingTx:   blkTx,
-					})
-
-					// Remove this ticket and continue with the next one.
-					delete(tickets, ticketHash)
-					continue outer
+		// Filter match means a ticket is likely spent in this block. Get the
+		// full block to confirm.
+		iBlock, err := dcrdClient.GetBlock(iHash)
+		if err != nil {
+			return nil, 0, err
+		}
+
+		for i, ticket := range tickets {
+			// The regular transaction tree does not need to be checked because
+			// tickets can only be spent by vote or revoke transactions which
+			// are always in the stake tree.
+			for _, blkTx := range iBlock.STransactions {
+				if !txSpendsTicket(blkTx, ticket.hash) {
+					continue
 				}
 
-				// Ticket is not spent in block.
+				// Confirmed - ticket is spent in block.
+
+				spent = append(spent, spentTicket{
+					dbTicket:     ticket.dbTicket,
+					expiryHeight: ticket.dbTicket.PurchaseHeight + int64(params.TicketMaturity) + int64(params.TicketExpiry),
+					heightSpent:  iHeight,
+					spendingTx:   blkTx,
+				})
+
+				// Remove this ticket and it's script before continuing with the
+				// next one.
+				tickets = nonOrderPreservingRemove(tickets, i)
+				scripts = nonOrderPreservingRemove(scripts, i)
```

Comment:
Don't forget to do `i--`.  You removed the current index which means everything else moved up one and thus the same index needed to be repeated.